### PR TITLE
PD-428: Implement IconButton Component

### PR DESCRIPTION
**Pull request by brookescarlett**
Screen Shot 2019-11-07 at 9 36 56 AM
Screen Shot 2019-11-07 at 9 37 00 AM

## ✍️ Proposed changes

- Initial implementation of IconButton component
- Will serve as a wrapper around Icons, providing the appearance, sizing, accessibility and interaction to make them buttons
- Precursor for MongoNav

🎟 _Jira ticket:_ [PD-428](https://jira.mongodb.org/browse/PD-428)

## 🛠 Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### CODE REVIEW

File: packages/icon-button/package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "main": "./dist/index.js",
+  "types": "./dist/index.d.ts",
+  "scripts": {
+    "build": "../../node_modules/.bin/webpack --config ../../webpack.config.js"
+  },
+  "license": "Apache-2.0",
+  "publishConfig": {
+    "access": "public"
+  },
+  "devDependencies": {
```
Comment (hswolff) [inline packages/icon-button/package.json]: Should be dependencies right?

File: packages/icon-button/src/IconButton.story.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  height: 60px;
+  background-color: ${uiColors.gray.dark3};
+  display: flex;
+  align-items: center;
+  justify-content: center;
+`;
+
+storiesOf('IconButton', module)
+  .add('Default', () => (
+    <IconButton
```
Comment (hswolff) [inline packages/icon-button/src/IconButton.story.tsx]: I was under the impression `IconButton` would shadow the available Icons that we have.

Also I just realized....the `glyph` prop on the `Icon` component  doesn't contain the list of available glyphs. I.e. it'd be better if we had an array of the available strings as the value for the glyph prop then we could autocomplete nicely.

And then this component would shadow that.

File: packages/icon-button/package.json (hunk 1 above)
Comment (brookescarlett) [reply packages/icon-button/package.json]: 👍

File: packages/icon-button/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+The event handler function for the 'onclick' event. Receives the associated `event` object as the first argument.
+
+### className
+
+**Type:** `string`
+
+**Default:** `''`
+
+Adds a className to the class attribute.
```
Comment (DesignerDave) [inline packages/icon-button/README.md]: Of which element?

File: packages/icon-button/src/IconButton.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { render, fireEvent, cleanup } from '@testing-library/react';
+import IconButton from './IconButton';
+
+afterAll(cleanup);
+
+describe('packages/icon-button', () => {
+  const onClick = jest.fn();
+  const className = 'test-icon-button-class';
+  const testId = 'test-icon-button-component';
+  const icon = `
```
Comment (DesignerDave) [inline packages/icon-button/src/IconButton.spec.tsx]: Won't this render a string, not an SVG?

File: packages/icon-button/src/IconButton.tsx [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+interface ButtonIconButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedIconButtonProps {
+  href?: null;
+}
+
+type IconButtonProps = LinkIconButtonProps | ButtonIconButtonProps;
+
+const removeButtonStyle = css`
```
Comment (DesignerDave) [inline packages/icon-button/src/IconButton.tsx]: Is there a reason we don't always apply these styles?

File: packages/icon-button/src/IconButton.tsx [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  ...rest
+}: IconButtonProps) {
+  const Root = href ? 'a' : 'button';
+
+  return (
+    <Root
+      {...rest}
+      onClick={disabled ? undefined : onClick}
+      href={href ? href : undefined}
+      aria-disabled={disabled ? true : false}
```
Comment (DesignerDave) [inline packages/icon-button/src/IconButton.tsx]: Disabled is a boolean, and appears to always be defined.

File: packages/icon-button/README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  <Icon glyph="Ellipsis" />
+</IconButton>;
+```
+
+**Output HTML**
+
+```html
+<button aria-disabled="false" class="leafygreen-ui-194rp0y">
+  <span class="leafygreen-ui-1rvdyoi"
+    ><svg width="16px" height="16px" viewBox="0 0 16 16" version="1.1">
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon-button/README.md]: Prettier does this whenever I save the file...

File: packages/icon-button/README.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      <title>Ellipsis</title>
+      <desc>Created with Sketch.</desc>
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon-button/README.md]: Is what? 😱

File: packages/icon-button/README.md (hunk 3 above)
Comment (brookescarlett) [reply packages/icon-button/README.md]: 👍 and reformatted to table

File: packages/icon-button/src/IconButton.spec.tsx (hunk 4 above)
Comment (brookescarlett) [reply packages/icon-button/src/IconButton.spec.tsx]: 👍

File: packages/icon-button/src/IconButton.tsx (hunk 5 above)
Comment (brookescarlett) [reply packages/icon-button/src/IconButton.tsx]: We can apply outside of a condition, no problem!

Comment (brookescarlett) [reply packages/icon-button/src/IconButton.tsx]: 👍

File: packages/icon-button/src/IconButton.tsx (hunk 6 above)
Comment (brookescarlett) [reply packages/icon-button/src/IconButton.tsx]: 👍

File: packages/icon-button/src/IconButton.story.tsx (hunk 2 above)
Comment (DesignerDave) [reply packages/icon-button/src/IconButton.story.tsx]: Hmm... We would need to update the `createIconComponent` method to accept an optional `Glyphs` type parameter that defaults to `string` for that. Can totally do that, but it doesn't yet support this.

Comment (brookescarlett) [reply packages/icon-button/src/IconButton.story.tsx]: AFAIK (<- haha love using those) `<Icon />` does support this autocomplete

File: packages/icon-button/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      <title>Ellipsis</title>
+      <g
+        id="Ellipsis-Copy"
+        stroke="none"
+        stroke-width="1"
+        fill="none"
+        fill-rule="evenodd"
+      >
+        <path
+          d="M2,6.5 C2.828125,6.5 3.5,7.171875 3.5,8 C3.5,8.828125 2.828125,9.5 2,9.5 C1.171875,9.5 0.5,8.828125 0.5,8 C0.5,7.171875 1.171875,6.5 2,6.5 Z M8,6.5 C8.828125,6.5 9.5,7.171875 9.5,8 C9.5,8.828125 8.828125,9.5 8,9.5 C7.171875,9.5 6.5,8.828125 6.5,8 C6.5,7.171875 7.171875,6.5 8,6.5 Z M14,6.5 C14.828125,6.5 15.5,7.171875 15.5,8 C15.5,8.828125 14.828125,9.5 14,9.5 C13.171875,9.5 12.5,8.828125 12.5,8 C12.5,7.171875 13.171875,6.5 14,6.5 Z"
```
Comment (DesignerDave) [inline packages/icon-button/README.md]: Left this comment also in the Logo PR. Can we just do `d="..."` for SVG path data?

File: packages/icon-button/src/IconButton.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+      <g
+        id="Ellipsis-Copy"
+        stroke="none"
+        strokeWidth="1"
+        fill="none"
+        fillRule="evenodd"
+      >
+        <path
+          d="M2,6.5 C2.828125,6.5 3.5,7.171875 3.5,8 C3.5,8.828125 2.828125,9.5 2,9.5 C1.171875,9.5 0.5,8.828125 0.5,8 C0.5,7.171875 1.171875,6.5 2,6.5 Z M8,6.5 C8.828125,6.5 9.5,7.171875 9.5,8 C9.5,8.828125 8.828125,9.5 8,9.5 C7.171875,9.5 6.5,8.828125 6.5,8 C6.5,7.171875 7.171875,6.5 8,6.5 Z M14,6.5 C14.828125,6.5 15.5,7.171875 15.5,8 C15.5,8.828125 14.828125,9.5 14,9.5 C13.171875,9.5 12.5,8.828125 12.5,8 C12.5,7.171875 13.171875,6.5 14,6.5 Z"
+          id=""
```
Comment (DesignerDave) [inline packages/icon-button/src/IconButton.spec.tsx]: Unsupported character?

File: packages/icon-button/README.md (hunk 9 above)
Comment (brookescarlett) [reply packages/icon-button/README.md]: 👍

File: packages/icon-button/src/IconButton.spec.tsx (hunk 10 above)
Comment (brookescarlett) [reply packages/icon-button/src/IconButton.spec.tsx]: 👍

**Pull request merged**